corehq/form_processor/management/commands/fix_edit_forms_attachments.py

```diff
@@ -0,0 +1,121 @@
+
+from io import open
+import csv342 as csv
+
+from corehq.sql_db.util import get_db_aliases_for_partitioned_query
+from couchforms.models import XFormDeprecated
+from corehq.form_processor.interfaces.dbaccessors import FormAccessors
+from corehq.form_processor.exceptions import XFormNotFound
+from corehq.form_processor.models import XFormOperationSQL
+from corehq.form_processor.backends.sql.dbaccessors import FormAccessorSQL
+from corehq.form_processor.backends.couch.dbaccessors import FormAccessorCouch
+from corehq.util.log import with_progress_bar
+
+
+class Command(BaseCommand):
+    def add_arguments(self, parser):
+        parser.add_argument('--source', help='sql/couch')
+        parser.add_argument('--update', action='store_true', help='actually update')
+        parser.add_argument('--inspect', action='store_true', help='just write findings to file')
+
+    @staticmethod
+    def _find_sql_forms_with_missing_attachments():
+        """
+        find all forms that are a result of a form edition
+        find the deprecated form and then compare attachments between them
+        :return:
+        """
+        add_forms_with_attachments = defaultdict(list)
+        form_accessor = FormAccessorSQL()
+        for db_alias in get_db_aliases_for_partitioned_query():
+            print('scanning %s' % db_alias)
+            edit_form_ids = (XFormOperationSQL.objects.using(db_alias)
+                             .filter(operation='edit')
+                             .values_list('form_id', flat=True))
+            for edit_form_id in with_progress_bar(edit_form_ids):
+                edited_form = form_accessor.get_form(edit_form_id)
+                deprecated_form_id = edited_form.deprecated_form_id
+                original_attachments = list(FormAccessorSQL.get_attachments_for_forms([deprecated_form_id]))
+                new_attachments = list(FormAccessorSQL.get_attachments_for_forms([edit_form_id]))
+                new_attachment_names = [a.name for a in new_attachments]
+                for attachment in original_attachments:
+                    if attachment.name not in new_attachment_names:
+                        assert attachment.form_id == deprecated_form_id
+                        add_forms_with_attachments[edit_form_id].append(
+                            (attachment.attachment_id, deprecated_form_id)
+                        )
+        return add_forms_with_attachments
+
+    @staticmethod
+    def _find_couch_forms_with_missing_attachments():
+        """
+        find deprecated form ids,
+        then look for the new form and compare attachments between them
+        :return:
+        """
+        add_forms_with_attachments = defaultdict(list)
+        deprecated_form_ids = [
+            f['id'] for f in XFormDeprecated.view(
+                'all_docs/by_doc_type', startkey=['XFormDeprecated'], endkey=['XFormDeprecated', {}],
+                include_docs=False, reduce=False).all()]
```

Comment:
@millerdev those are 33269 forms. Should be fine.

@orangejenny 
considering we are seeing forms from even before that date, i am thinking if
1. the logic to get these forms is incorrect because this should have happened only after May 15?
2. There are some genuine cases where an attachment was removed and we would now add it back? I am not sure if edit forms also allowed to remove attachments or if other cases before May 15 genuinely removed attachments from the forms?

Reply:
i think this is related to couch-sql migration. that probably uses the same process of deprecating couch form and moving it to sql and what we are seeing here is some issue with that process. But i am not sure.

Reply:
Couch-sql migration doesn't do any form deprecation (checked with @jmtroth0 ).

> There are some genuine cases where an attachment was removed and we would now add it back?

I'm not especially worried about having purposefully removed an attachment in the past, because I don't think we've given users a way to remove attachments - I'd guess this whole issue is actually consistent with how old edit forms worked (since you couldn't update multimedia from web apps), it's just less obvious to the user what's going on (since with old edit forms, you'd see the multimedia question and a warning).